src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#[deny(clippy::all)]
+///! Graph API Traits
+///
+use rand::SeedableRng;
+
+#[allow(dead_code)]
+/// A graph layer name.
+struct Layer(&'static str);
+
+/// Abstract object in a graph, eg. node or edge.
+trait GraphObject {
+    /// Identifier of all graph objects.
+    type Id;
+
+    /// Local graph object data.
+    type Data;
+
+    /// Return the edge id.
+    fn id(&self) -> Self::Id;
+
+    /// Get local data.
+    fn data(&self) -> &Self::Data;
+
+    /// Get mutable local data.
+    fn data_mut(&self) -> &mut Self::Data;
+}
+
+/// A graph  node.
+trait Node: GraphObject {
+    /// The edge type connected to this node.
+    type Edge: GraphObject;
+
+    /// Get connected nodes.
+    fn neighbors(&self) -> Iterator<Item = &Self>;
+
+    /// Get incoming and outgoing edges.
+    fn edges(&self) -> Iterator<Item = &Self::Edge>;
+}
+
+/// A graph edge between two nodes.
+trait Edge: GraphObject {
+    /// Get the edge weight.
+    fn weight(&self) -> f64;
```

Comment:
Why isn't the weight a type parameter or part of the "bag of data" each `GraphObject` hold? Having an `f64` here makes testing problematic for us, because in tests (where we don't care about performance) we model a weight as a fraction.

Reply:
Because this guarantees that an edge has a weight - otherwise you have to deal with `Option` everywhere you deal with weights. With regards to testing, I don't see a problem with having the weight type be parametrized: `fn weight(&self) -> Self::Weight`.

Reply:
Yes, this is something I have also had to come to terms with. In our case, I have solved this with an implementation-specific `get_weight` callback:

https://github.com/oscoin/osrank-rs/blob/master/src/algorithm.rs#L94

I agree though that it's not as nice as having the `weight` function as part of the trait itself, at the cost of an extra layer of indirection. In the osrank implementation, we don't have the extra `Node` and `Edge` traits. Pick your poison 😬